go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1294,27 +1294,55 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
 	if err == nil {
 		return nil
 	}
-	err = tsv.convertError(ctx, sql, bindVariables, err)
 
-	if logStats != nil {
-		logStats.Error = err
-	}
-	errCode := vterrors.Code(err)
+	errStr := err.Error()
+	errCode := tsv.convertErrorCode(err)
 	tabletenv.ErrorStats.Add(errCode.String(), 1)
 
-	logMethod := log.Errorf
+	logMethod := tabletenv.Errorf
 	// Suppress or demote some errors in logs.
 	switch errCode {
 	case vtrpcpb.Code_FAILED_PRECONDITION, vtrpcpb.Code_ALREADY_EXISTS:
-		return err
+		logMethod = nil
 	case vtrpcpb.Code_RESOURCE_EXHAUSTED:
 		logMethod = logTxPoolFull.Errorf
 	case vtrpcpb.Code_ABORTED:
-		logMethod = log.Warningf
+		logMethod = tabletenv.Warningf
 	case vtrpcpb.Code_INVALID_ARGUMENT, vtrpcpb.Code_DEADLINE_EXCEEDED:
-		logMethod = log.Infof
+		logMethod = tabletenv.Infof
+	}
+
+	origErr := err
+	err = formatErrorWithCallerID(ctx, vterrors.New(errCode, errStr))
+	if logMethod != nil {
+		logMethod("%v: %v", err, queryAsString(sql, bindVariables))
+	}
+
+	// If TerseErrors is on, strip the error message returned by MySQL and only
+	// keep the error number and sql state.
+	// We assume that bind variable have PII, which are included in the MySQL
+	// query and come back as part of the error message. Removing the MySQL
+	// error helps us avoid leaking PII.
+	// There are two exceptions:
+	// 1. If no bind vars were specified, it's likely that the query was issued
+	// by someone manually. So, we don't suppress the error.
+	// 2. FAILED_PRECONDITION errors. These are caused when a failover is in progress.
+	// If so, we don't want to suppress the error. This will allow VTGate to
+	// detect and perform buffering during failovers.
+	if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+		sqlErr, ok := origErr.(*mysql.SQLError)
+		if ok {
+			sqlState := sqlErr.SQLState()
+			errnum := sqlErr.Number()
+			errStr = fmt.Sprintf("(errno %d) (sqlstate %s) during query: %s", errnum, sqlState, sqlparser.TruncateForLog(sql))
+			err = formatErrorWithCallerID(ctx, vterrors.New(errCode, errStr))
+		}
 	}
-	logMethod("%v: %v", err, queryAsString(sql, bindVariables))
+
+	if logStats != nil {
```

Comment:
Tagging @alainjobart / @michael-berlin 
I think this is an acceptable trade-off, to not have the full error in the log stats. In fact, it may even be preferable because I don't think we tag this field as containing PII. If necessary, the error logs will still have the full error.

Reply:
This _should_ be the same behavior as it was before -- note that in the previous version we assigned the `err` to log stats after calling `convertError()` on it, which was where the redaction happened previously.

So both before and after this change. log stats does not have the full error.

Reply:
Oh. Of course! I'll merge once the tests pass.